src/middle/Utils.ml

```diff
@@ -87,3 +79,42 @@ let%expect_test "all but last n" =
   let l = all_but_last_n [1; 2; 3; 4] 2 in
   print_s [%sexp (l : int list)] ;
   [%expect {| (1 2) |}]
+
+(* Utilities for using Tuples and Transformations together *)
+
+let zip_tuple_trans_exn pst trans =
+  let tms =
+    match trans with
+    | Transformation.TupleTransformation tms -> tms
+    | _ ->
+        Common.FatalError.fatal_error_msg
+          [%message "Internal error: expected TupleTransformation with Tuple"]
+  in
+  let rec tuple_psts pst =
+    match pst with
+    | Type.Unsized (UTuple uts) -> List.map ~f:(fun ut -> Type.Unsized ut) uts
+    | Type.Unsized (UArray ut) -> tuple_psts (Type.Unsized ut)
+    | Type.Sized (STuple sts) -> List.map ~f:(fun st -> Type.Sized st) sts
+    | Type.Sized (SArray (st, _)) -> tuple_psts (Type.Sized st)
+    | _ ->
+        Common.FatalError.fatal_error_msg
```

Comment:
Rather than taking a `trans` that _must_ be a `TupleTransformation tms` this function should take the list `tms` directly. Many call-sites already have already `match`ed the transformation and those that haven't should acknowledge the potential failure mode. Also, `zip_tuple_trans_exn` is not used outside of `Utils`; it seems to exist just for code sharing between `zip_stuple_trans_exn` and `zip_utuple_trans_exn`. However, there's not that much common code and those unreachable `"Internal error in zip_tuple"` errors add confusion. I don't think it's worth the trouble.

Reply:
The benefit of this function existing is exactly because we often have matched on either the type or the transformation but rarely both. If we had both already, we wouldn't need much beyond a zip, so the utility of this is that it lets us keep the somewhat ugly `fatal_error` logic out of most of these places.

Unlike the TupleAD stuff which it is up to us to constantly book-keep, the transformations come pretty directly from parsing and the failure branches are much closer to dead code.

Reply:
I disagree.

`zip_utuple_trans` is used only once, in the typecheker right after matching on the transformation.
`zip_stuple_trans` is used six times:
* in `Pretty_printing` right after matching on the transformation
* in `Transform_mir` where the code matches on type but could just as well match on the transformation
* three times in `Ast_to_mir`, two of which already matched on the transformation
* in `Generate_debug_data`

So by my count all but two callsites can pass the transformation list directly without wrapping it in a `TupleTransformation`.
If you want a helper for the remaining two I think the most natural one would be a `tuple_trans_exn : 'e t -> 'e t list` in `Transformation`.

Reply:
I made the changes. The callsites where we were already matching on TupleTransformation definitely seem nicer, but the other ones are just as bad if not worse IMO. 
I left the `_exn` suffix on the functions since they could still fail if not passed a tuple type.